vyper/contracts/libs/AmountDeriver.vy

```diff
@@ -0,0 +1,144 @@
+# @version 0.3.3
+
+"""
+@dev  A struct containing the data used to apply a
+      fraction to an order.
+"""
+struct FractionData:
+    numerator: uint256  # The portion of the order that should be filled.
+    denominator: uint256  # The total size of the order
+    offererConduitKey: bytes32  # The offerer's conduit key.
+    fulfillerConduitKey: bytes32  # The fulfiller's conduit key.
+    duration: uint256  # The total duration of the order.
+    elapsed: uint256  # The time elapsed since the order's start time.
+    remaining: uint256  # The time left until the order's end time.
+
+@internal
+@pure
+def _locateCurrentAmount(
+    startAmount: uint256,
+    endAmount: uint256,
+    elapsed: uint256,
+    remaining: uint256,
+    duration: uint256,
+    roundUp: bool
+) -> uint256:
+    """
+    @dev Internal pure function to derive the current amount of a given item
+        based on the current price, the starting price, and the ending
+        price. If the start and end prices differ, the current price will be
+        extrapolated on a linear basis.
+
+    @param startAmount The starting amount of the item.
+    @param endAmount   The ending amount of the item.
+    @param elapsed     The time elapsed since the order's start time.
+    @param remaining   The time left until the order's end time.
+    @param duration    The total duration of the order.
+    @param roundUp     A boolean indicating whether the resultant amount
+                       should be rounded up or down.
+
+    @return The current amount.
+    """
+    # Only modify end amount if it doesn't already equal start amount.
+    if startAmount != endAmount:
+        # Leave extra amount to add for rounding at zero (i.e. round down).
+        extraCeiling: uint256 = 0
+
+        # If rounding up, set rounding factor to one less than denominator.
+        if roundUp:
+            extraCeiling = duration - 1
+
+        # Aggregate new amounts weighted by time with rounding factor.
+        totalBeforeDivision: uint256 = ((startAmount * remaining) +
+            (endAmount * elapsed) +
+            extraCeiling)
+
+        # Division is performed without zero check as it cannot be zero.
+        newAmount: uint256 = totalBeforeDivision / duration
+
+        # Return the current amount (expressed as endAmount internally).
+        return newAmount
+
+    # Return the original amount (now expressed as endAmount internally).
+    return endAmount
```

Comment:
nit: stylistically, might be good to fast path out `if startAmount == endAmount`. then the more complicated part can be less indented.

Reply:
I also think `# Division is performed without zero check as it cannot be zero.` comment is maybe not true? Not sure whether vyper does a zero check on division, guessing it does though

Reply:
> I also think `# Division is performed without zero check as it cannot be zero.` comment is maybe not true? Not sure whether vyper does a zero check on division, guessing it does though

vyper division does do a 0 check. if you wanted to really skip the zero check, could use `unsafe_div(a, b)`

Reply:
I need to be more careful with the comments than just copying from the reference :) 
Will go with `unsafe_div`.